autogen/oai/gemini.py

```diff
@@ -0,0 +1,288 @@
+from openai.types.chat import ChatCompletion
+from openai.types.chat.chat_completion import ChatCompletionMessage, Choice
+from openai.types.completion_usage import CompletionUsage
+from PIL import Image
+from proto.marshal.collections.repeated import RepeatedComposite
+from pydash import max_
+from typing_extensions import Self, override
+
+# from autogen.agentchat.contrib.img_utils import _to_pil, get_image_data
+from autogen.token_count_utils import count_token
+
+
+class GeminiClient:
+    """
+    _summary_
+
+    _extended_summary_
+
+    TODO: this Gemini implementation does not support the following features yet:
+    - function_call (OpenAI)
+    - tool_calls (OpenAI)
+    - safety_setting (Gemini)
+    - Multi-turn chat for vision model (Gemini)
+    - multiple responses at the same time (Gemini)
+    """
+
+    def __init__(self, **kwargs):
+        self.api_key = kwargs.get("api_key", None)
+        if self.api_key is None:
+            self.api_key = os.getenv("GOOGLE_API_KEY")
+
+        assert self.api_key is not None, (
+            "Please provide api_key in OAI_CONFIG_LIST " "or set the GOOGLE_API_KEY env variable."
+        )
+
+    def call(self, params: Dict) -> ChatCompletion:
+        model_name = params.get("model", "gemini-pro")
+        params.get("api_type", "google")  # not used
+        messages = params.get("messages", [])
+        stream = params.get("stream", False)
+        n_response = params.get("n", 1)
+        params.get("temperature", 0.5)
+        params.get("top_p", 1.0)
+        params.get("max_tokens", 1024)
+        # TODO: handle these parameters in GenerationConfig
+
+        if stream:
+            # TODO: support streaming
+            # warn user that streaming is not supported
+            print("Streaming is not supported for Gemini yet. Please set stream=False.")
+
+        if n_response > 1:
+            print("Gemini only supports `n=1` for now. We only generate one response.")
+
+        if "vision" not in model_name:
+            # A. create and call the chat model.
+            gemini_messages = oai_messages_to_gemini_messages(messages)
+
+            # we use chat model by default
+            model = genai.GenerativeModel(model_name)
+            genai.configure(api_key=self.api_key)
+            chat = model.start_chat(history=gemini_messages[:-1])
+            try:
+                response = chat.send_message(gemini_messages[-1].parts[0].text, stream=stream)
+            except InternalServerError as e:
+                print(e)
+                print("InternalServerError `500` occurs when calling Gemini's chat model. Retry in 5 seconds...")
+                time.sleep(5)
+                return self.call(params)
+            except Exception as e:
+                print("Exception occurred while calling Gemini API:", e)
+                ans = "TERMINATE"
+            else:
+                # ans = response.text # failed. Not sure why.
```

Comment:
Response can fail sometimes if the answer is "complex" for example using of functions (which is if I'm not wrong is in beta of gemini)